trmm perf test nightly build error: clang/8+cuda tpl build

The following nightly test failed on kokkos-dev-2 after merge of PR #720, @e10harvey can you take a look? From the compiler message looks like either a routine is missing a `KOKKOS_INLINE_FUNCTION` marking, or a host function is being incorrectly called from a kernel in a `Kokkos::parallel_for`

clang-8.0-Cuda_OpenMP-release (build failed)
https://jenkins-son.sandia.gov/job/KokkosKernels_KokkosDev2_SPOTCHECK_TPLS/83/console

Other impacted builds:
https://jenkins-son.sandia.gov/view/KokkosKernels/job/KokkosKernels_KokkosDev2_CLANG8_CUDA10_cpp17/
https://jenkins-son.sandia.gov/view/KokkosKernels/job/KokkosKernels_KokkosDev2_CLANG8_CUDA10_cpp17/
https://jenkins-son.sandia.gov/view/KokkosKernels/job/KokkosKernels_KokkosDev_CLANG7_CUDA9/84/console

**First compile error**
```bash
In file included from /home/jenkins/slave/workspace/KokkosKernels_KokkosDev2_SPOTCHECK_TPLS/kokkos-kernels/perf_test/blas/blas3/KokkosBlas3_perf_test.cpp:45:
/home/jenkins/slave/workspace/KokkosKernels_KokkosDev2_SPOTCHECK_TPLS/kokkos-kernels/perf_test/blas/blas3/KokkosBlas3_trmm_perf_test.hpp:303:17: error: reference to __host__ function 'trmm<Kokkos::View<double **, Kokkos::LayoutStride, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<0> >, Kokkos::View<double **, Kokkos::LayoutStride, Kokkos::Device<Kokkos::Cuda, Kokkos::CudaSpace>, Kokkos::MemoryTraits<0> > >' in __host__ __device__ function
 KokkosBlas::trmm(&trmm_args_.side, &trmm_args_.uplo, &trmm_args_.trans,
```

**Reproducer instructions:**
```bash
   #   Load modules:
         source /projects/sems/modulefiles/utils/sems-modules-init.sh ; module use /home/projects/x86-64/modulefiles/local
         module purge
         module load sems-env sems-cmake/3.12.2 kokkos-env clang/8.0 cuda/10.0
 
       $KOKKOSKERNELS_PATH/cm_generate_makefile.bash --with-devices=Cuda,OpenMP --arch=SNB,Volta70 --compiler=/home/projects/x86-64/clang/8.0/bin/clang++ --cxxflags="-O3 -Wall -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized -Wno-pass-failed " --cxxstandard="11" --ldflags="" --with-cuda=/home/projects/x86-64/cuda/10.0 --kokkos-path=$KOKKOS_PATH --kokkoskernels-path=$KOKKOSKERNELS_PATH --with-scalars='double,complex_double' --with-ordinals=int --with-offsets=int,size_t --with-layouts=LayoutLeft --with-tpls=blas    --with-options= --with-cuda-options= --no-examples 
```

Edit: Added links to other affected builds

--- a/perf_test/blas/blas/KokkosBlas_perf_test.cpp
+++ b/perf_test/blas/blas/KokkosBlas_perf_test.cpp
@@ -247,7 +247,9 @@ int main(int argc, char **argv) {
         out_file         = optarg;
         options.out_file = std::string(out_file);
         break;
-      case 'r': options.blas_routines = std::string(optarg);
+      case 'r': 
+        options.blas_routines = std::string(optarg);
+        break;
       case '?':
       default: __blas_perf_test_input_error(argv, option_idx);
     }

--- a/perf_test/blas/blas/KokkosBlas_trtri_perf_test.hpp
+++ b/perf_test/blas/blas/KokkosBlas_trtri_perf_test.hpp
@@ -201,6 +201,7 @@ void __do_trtri_serial_batched(options_t options, trtri_args_t trtri_args) {
   return;
 }
 
+#if !defined(KOKKOS_ENABLE_CUDA)
 template <class ExecutionSpace>
 struct parallel_blas_trtri {
   trtri_args_t trtri_args_;
@@ -214,9 +215,11 @@ struct parallel_blas_trtri {
     KokkosBlas::trtri(&trtri_args_.uplo, &trtri_args_.diag, svA);
   }
 };
+#endif // !KOKKOS_ENABLE_CUDA
 
 template <class scalar_type, class vta, class device_type>
 void __do_trtri_parallel_blas(options_t options, trtri_args_t trtri_args) {
+  #if !defined(KOKKOS_ENABLE_CUDA)
   uint32_t warm_up_n = options.warm_up_n;
   uint32_t n         = options.n;
   Kokkos::Timer timer;
@@ -237,6 +240,12 @@ void __do_trtri_parallel_blas(options_t options, trtri_args_t trtri_args) {
                        parallel_blas_trtri_functor);
   Kokkos::fence();
   __trtri_output_csv_row(options, trtri_args, timer.seconds());
+  #else
+  std::cerr << std::string(__func__) << 
+    " disabled since KOKKOS_ENABLE_CUDA is defined." << 
+    std::endl;
+  __trtri_output_csv_row(options, trtri_args, -1);
+  #endif // !KOKKOS_ENABLE_CUDA
   return;
 }
 

--- a/perf_test/blas/blas/KokkosBlas_trtri_perf_test.hpp
+++ b/perf_test/blas/blas/KokkosBlas_trtri_perf_test.hpp
@@ -329,19 +329,23 @@ trtri_args_t __do_setup(options_t options, matrix_dims_t dim) {
   uint64_t seed = Kokkos::Impl::clock_tic();
   Kokkos::Random_XorShift64_Pool<execution_space> rand_pool(seed);
   decltype(dim.a.m) min_dim = dim.a.m < dim.a.n ? dim.a.m : dim.a.n;
+  typename vta::HostMirror host_A;
   STATUS;
 
   trtri_args.uplo = options.blas_args.trtri.trtri_args.c_str()[0];
   trtri_args.diag = options.blas_args.trtri.trtri_args.c_str()[1];
   trtri_args.A    = vta("trtri_args.A", options.n, dim.a.m, dim.a.n);
+  host_A = Kokkos::create_mirror_view(trtri_args.A);
 
   Kokkos::fill_random(trtri_args.A, rand_pool,
                       Kokkos::rand<Kokkos::Random_XorShift64<execution_space>,
                                    scalar_type>::max());
+  Kokkos::deep_copy(host_A, trtri_args.A);
+
   if (trtri_args.uplo == 'U' || trtri_args.uplo == 'u') {
     // Make A upper triangular
     for (uint32_t k = 0; k < options.n; ++k) {
-      auto A = Kokkos::subview(trtri_args.A, k, Kokkos::ALL(), Kokkos::ALL());
+      auto A = Kokkos::subview(host_A, k, Kokkos::ALL(), Kokkos::ALL());
       for (int i = 1; i < dim.a.m; i++) {
         for (int j = 0; j < i; j++) {
           A(i, j) = scalar_type(0);
@@ -353,7 +357,7 @@ trtri_args_t __do_setup(options_t options, matrix_dims_t dim) {
     // Kokkos::parallel_for("toLowerLoop", options.n, KOKKOS_LAMBDA (const int&
     // i) {
     for (uint32_t k = 0; k < options.n; ++k) {
-      auto A = Kokkos::subview(trtri_args.A, k, Kokkos::ALL(), Kokkos::ALL());
+      auto A = Kokkos::subview(host_A, k, Kokkos::ALL(), Kokkos::ALL());
       for (int i = 0; i < dim.a.m - 1; i++) {
         for (int j = i + 1; j < dim.a.n; j++) {
           A(i, j) = scalar_type(0);
@@ -364,12 +368,14 @@ trtri_args_t __do_setup(options_t options, matrix_dims_t dim) {
 
   if (trtri_args.diag == 'U' || trtri_args.diag == 'u') {
     for (uint32_t k = 0; k < options.n; ++k) {
-      auto A = Kokkos::subview(trtri_args.A, k, Kokkos::ALL(), Kokkos::ALL());
+      auto A = Kokkos::subview(host_A, k, Kokkos::ALL(), Kokkos::ALL());
       for (int i = 0; i < min_dim; i++) {
         A(i, i) = scalar_type(1);
       }
     }
   }
+
+  Kokkos::deep_copy(trtri_args.A, host_A);
 
   return trtri_args;
 }

--- a/perf_test/blas/blas3/KokkosBlas3_perf_test.cpp
+++ b/perf_test/blas/blas3/KokkosBlas3_perf_test.cpp
@@ -259,7 +259,9 @@ int main(int argc, char **argv) {
         out_file         = optarg;
         options.out_file = std::string(out_file);
         break;
-      case 'r': options.blas_routines = std::string(optarg);
+      case 'r': 
+        options.blas_routines = std::string(optarg);
+        break;
       case '?':
       default: __blas3_perf_test_input_error(argv, option_idx);
     }

--- a/perf_test/blas/blas3/KokkosBlas3_trmm_perf_test.hpp
+++ b/perf_test/blas/blas3/KokkosBlas3_trmm_perf_test.hpp
@@ -289,6 +289,7 @@ void __do_trmm_serial_batched(options_t options, trmm_args_t trmm_args) {
   return;
 }
 
+#if !defined(KOKKOS_ENABLE_CUDA)
 template <class ExecutionSpace>
 struct parallel_blas_trmm {
   trmm_args_t trmm_args_;
@@ -304,9 +305,11 @@ struct parallel_blas_trmm {
                      &trmm_args_.diag, trmm_args_.alpha, svA, svB);
   }
 };
+#endif // !KOKKOS_ENABLE_CUDA
 
 template <class scalar_type, class vta, class vtb, class device_type>
 void __do_trmm_parallel_blas(options_t options, trmm_args_t trmm_args) {
+  #if !defined(KOKKOS_ENABLE_CUDA)
   uint32_t warm_up_n = options.warm_up_n;
   uint32_t n         = options.n;
   Kokkos::Timer timer;
@@ -315,6 +318,7 @@ void __do_trmm_parallel_blas(options_t options, trmm_args_t trmm_args) {
   functor_type parallel_blas_trmm_functor(trmm_args);
 
   STATUS;
+
 
   Kokkos::parallel_for("parallelBlasWarmUpLoopTrmm",
                        Kokkos::RangePolicy<execution_space>(0, warm_up_n),
@@ -327,6 +331,12 @@ void __do_trmm_parallel_blas(options_t options, trmm_args_t trmm_args) {
                        parallel_blas_trmm_functor);
   Kokkos::fence();
   __trmm_output_csv_row(options, trmm_args, timer.seconds());
+  #else
+  std::cerr << std::string(__func__) << 
+    " disabled since KOKKOS_ENABLE_CUDA is defined." << 
+    std::endl;
+  __trmm_output_csv_row(options, trmm_args, -1);
+  #endif // !KOKKOS_ENABLE_CUDA
   return;
 }
 

--- a/perf_test/blas/blas3/KokkosBlas3_trmm_perf_test.hpp
+++ b/perf_test/blas/blas3/KokkosBlas3_trmm_perf_test.hpp
@@ -494,6 +494,7 @@ trmm_args_t __do_setup(options_t options, matrix_dims_t dim) {
   uint64_t seed = Kokkos::Impl::clock_tic();
   Kokkos::Random_XorShift64_Pool<execution_space> rand_pool(seed);
   decltype(dim.a.m) min_dim = dim.a.m < dim.a.n ? dim.a.m : dim.a.n;
+  typename vta::HostMirror host_A;
   STATUS;
 
   trmm_args.side  = options.blas_args.trmm.trmm_args.c_str()[0];
@@ -502,14 +503,17 @@ trmm_args_t __do_setup(options_t options, matrix_dims_t dim) {
   trmm_args.diag  = options.blas_args.trmm.trmm_args.c_str()[3];
   trmm_args.A     = vta("trmm_args.A", options.n, dim.a.m, dim.a.n);
   trmm_args.B     = vtb("trmm_args.B", options.n, dim.b.m, dim.b.n);
+  host_A = Kokkos::create_mirror_view(trmm_args.A);
 
   Kokkos::fill_random(trmm_args.A, rand_pool,
                       Kokkos::rand<Kokkos::Random_XorShift64<execution_space>,
                                    scalar_type>::max());
-  if (trmm_args.uplo == 'U' || trmm_args.uplo == 'u') {
+  Kokkos::deep_copy(host_A, trmm_args.A);
+
+  if (trmm_args.uplo == 'U' || trmm_args.uplo == 'u') {  
     // Make A upper triangular
     for (uint32_t k = 0; k < options.n; ++k) {
-      auto A = Kokkos::subview(trmm_args.A, k, Kokkos::ALL(), Kokkos::ALL());
+      auto A = Kokkos::subview(host_A, k, Kokkos::ALL(), Kokkos::ALL());
       for (int i = 1; i < dim.a.m; i++) {
         for (int j = 0; j < i; j++) {
           A(i, j) = scalar_type(0);
@@ -521,7 +525,7 @@ trmm_args_t __do_setup(options_t options, matrix_dims_t dim) {
     // Kokkos::parallel_for("toLowerLoop", options.n, KOKKOS_LAMBDA (const int&
     // i) {
     for (uint32_t k = 0; k < options.n; ++k) {
-      auto A = Kokkos::subview(trmm_args.A, k, Kokkos::ALL(), Kokkos::ALL());
+      auto A = Kokkos::subview(host_A, k, Kokkos::ALL(), Kokkos::ALL());
       for (int i = 0; i < dim.a.m - 1; i++) {
         for (int j = i + 1; j < dim.a.n; j++) {
           A(i, j) = scalar_type(0);
@@ -532,12 +536,13 @@ trmm_args_t __do_setup(options_t options, matrix_dims_t dim) {
 
   if (trmm_args.diag == 'U' || trmm_args.diag == 'u') {
     for (uint32_t k = 0; k < options.n; ++k) {
-      auto A = Kokkos::subview(trmm_args.A, k, Kokkos::ALL(), Kokkos::ALL());
+      auto A = Kokkos::subview(host_A, k, Kokkos::ALL(), Kokkos::ALL());
       for (int i = 0; i < min_dim; i++) {
         A(i, i) = scalar_type(1);
       }
     }
   }
+  Kokkos::deep_copy(trmm_args.A, host_A);
 
   Kokkos::fill_random(trmm_args.B, rand_pool,
                       Kokkos::rand<Kokkos::Random_XorShift64<execution_space>,

